7.Multiparty-UICollectionView/7.Multiparty-UICollectionView/ChatViewController.swift

```diff
@@ -0,0 +1,188 @@
+    }
+
+    /**
+     * Sets up an instance of OTPublisher to use with this session. OTPubilsher
+     * binds to the device camera and microphone, and will provide A/V streams
+     * to the OpenTok session.
+     */
+    fileprivate func doPublish() {
+        var error: OTError?
+        defer {
+            processError(error)
+        }
+        session.publish(publisher, error: &error)
+        
+        collectionView?.reloadData()
+    }
+    
+    /**
+     * Instantiates a subscriber for the given stream and asynchronously begins the
+     * process to begin receiving A/V content for this stream. Unlike doPublish,
+     * this method does not add the subscriber to the view hierarchy. Instead, we
+     * add the subscriber only after it has connected and begins receiving data.
+     */
+    fileprivate func doSubscribe(_ stream: OTStream) {
+        var error: OTError?
+        defer {
+            processError(error)
+        }
+        guard let subscriber = OTSubscriber(stream: stream, delegate: self)
+            else {
+                print("Error while subscribing")
+                return
+        }
+        session.subscribe(subscriber, error: &error)
+        subscribers.append(subscriber)        
+        collectionView?.reloadData()
+    }
+    
+    fileprivate func cleanupSubscriber(_ stream: OTStream) {
+        subscribers = subscribers.filter { $0.stream?.streamId != stream.streamId }
+        collectionView?.reloadData()
+    }
+    
+    fileprivate func processError(_ error: OTError?) {
+        if let err = error {
+            showAlert(errorStr: err.localizedDescription)
+        }
+    }
+    
+    fileprivate func showAlert(errorStr err: String) {
+        DispatchQueue.main.async {
+            let controller = UIAlertController(title: "Error", message: err, preferredStyle: .alert)
+            controller.addAction(UIAlertAction(title: "Ok", style: .default, handler: nil))
+            self.present(controller, animated: true, completion: nil)
+        }
+    }
+    
+    // MARK: - UICollectionView methods
+    override func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
+        return subscribers.count + 1
+    }
+    
+    override func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "videoCell", for: indexPath)
+        let videoView: UIView = {
+            if (indexPath.row == 0) {
+                return publisher.view!
+            } else {
+                let sub = subscribers[indexPath.row - 1]
+                return sub.view!
+            }
+        }()
+        
+        videoView.frame = cell.bounds
+        cell.addSubview(videoView)
+        return cell
+    }
+    
+    override func numberOfSections(in collectionView: UICollectionView) -> Int {
+        return 1
+    }
```

Comment:
This delegate can be removed without affecting the functionality if we want to do that 😃